File: golang/log.go
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package golang

import (
	"log"
	"os"
	"strconv"
	"strings"

	"github.com/apache/rocketmq-clients/golang/pkg/utils"
	"github.com/natefinch/lumberjack"
	"go.uber.org/zap"
	"go.uber.org/zap/zapcore"
)

const (
	CLIENT_LOG_ROOT     = "rocketmq.client.logRoot"
	CLIENT_LOG_MAXINDEX = "rocketmq.client.logFileMaxIndex"
	CLIENT_LOG_FILESIZE = "rocketmq.client.logFileMaxSize"
	CLIENT_LOG_LEVEL    = "rocketmq.client.logLevel"
	// CLIENT_LOG_ADDITIVE        = "rocketmq.client.log.additive"
	CLIENT_LOG_FILENAME = "rocketmq.client.logFileName"
	// CLIENT_LOG_ASYNC_QUEUESIZE = "rocketmq.client.logAsyncQueueSize"
	ENABLE_CONSOLE_APPENDER = "mq.consoleAppender.enabled"
)

var sugarBaseLogger *zap.SugaredLogger

func ResetLogger() {
	InitLogger()
}

func InitLogger() {
	writeSyncer := getLogWriter()
	isStdOut := utils.GetenvWithDef(ENABLE_CONSOLE_APPENDER, "false")
	if isStdOut == "true" {
		writeSyncer = os.Stdout
	}
	encoder := getEncoder()

	var atomicLevel = zap.NewAtomicLevel()
	switch strings.ToLower(os.Getenv(CLIENT_LOG_LEVEL)) {
	case "debug":
		atomicLevel.SetLevel(zap.DebugLevel)
	case "warn":
		atomicLevel.SetLevel(zap.WarnLevel)
	case "error":
		atomicLevel.SetLevel(zap.ErrorLevel)
	default:
		atomicLevel.SetLevel(zap.InfoLevel)
	}

	core := zapcore.NewCore(encoder, writeSyncer, atomicLevel)

	logger := zap.New(core, zap.AddCaller())
	sugarBaseLogger = logger.Sugar()
}

func getEncoder() zapcore.Encoder {
	encoderConfig := zap.NewProductionEncoderConfig()
	encoderConfig.EncodeTime = zapcore.ISO8601TimeEncoder
	encoderConfig.EncodeLevel = zapcore.CapitalLevelEncoder
	return zapcore.NewConsoleEncoder(encoderConfig)
}

func getLogWriter() zapcore.WriteSyncer {
	clientLogRoot := utils.GetenvWithDef(CLIENT_LOG_ROOT, os.Getenv("user.home")+"/logs/rocketmqlogs")
	clientLogMaxIndex := utils.GetenvWithDef(CLIENT_LOG_MAXINDEX, "10")
	clientLogFileName := utils.GetenvWithDef(CLIENT_LOG_FILENAME, "rocketmq_client_go.log")
	clientLogMaxFileSize := utils.GetenvWithDef(CLIENT_LOG_FILESIZE, "1073741824")

	logFileName := clientLogRoot + "/" + clientLogFileName
	maxFileIndex, err := strconv.Atoi(clientLogMaxIndex)
	if err != nil {
		log.Printf("%s='%s' is invalid and has been replaced with the default value %s", CLIENT_LOG_MAXINDEX, clientLogMaxIndex, "10")
		maxFileIndex = 10
	}
	maxFileSize, err := strconv.Atoi(clientLogMaxFileSize)
	if err != nil {
		log.Printf("%s='%s' is invalid and has been replaced with the default value %s", CLIENT_LOG_FILESIZE, clientLogMaxFileSize, "1073741824")
		maxFileSize = 1073741824
	}
	lumberJackLogger := &lumberjack.Logger{
		Filename:   logFileName,
		MaxSize:    maxFileSize,
		MaxBackups: maxFileIndex,
		Compress:   false,
	}
	return zapcore.AddSync(lumberJackLogger)
}

func init() {
	InitLogger()
}

File: rust/src/log.rs
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
use std::fs;
use std::fs::{File, OpenOptions};
use std::path::{PathBuf};

use slog::{o, Drain, Logger};
use slog_async::OverflowStrategy;

use crate::conf::LoggingFormat;

pub(crate) fn terminal_logger() -> Logger {
    let decorator = slog_term::TermDecorator::new().build();
    let drain = slog_term::FullFormat::new(decorator)
        .use_file_location()
        .build()
        .fuse();
    let drain = slog_async::Async::new(drain)
        .overflow_strategy(OverflowStrategy::Block)
        .chan_size(1)
        .build()
        .fuse();
    Logger::root(drain, o!())
}

fn json_logger(filepath: &Option<String>) -> Logger {
    // 设置默认值和解析传入的参数
    let fp = match filepath {
        None => {
            PathBuf::from("logs/rocketmq.client.log")
        }
        Some(p) => {
            PathBuf::from(p)
        }
    };
    // first check the director or file is existing?
    let file = if !fp.exists() {
        let create_file = |fp: &PathBuf| -> Option<File> {
            match OpenOptions::new()
                .create(true)
                .write(true)
                .append(true)
                .open(fp) {
                Ok(f) => Some(f),
                Err(_) => None
            }
        };
        // have the parent dir
        match fp.parent() {
            // or no have
            None => create_file(&fp),
            // have the parent dir, create it, if success create the file.
            Some(p) => match p.exists() {
                true => create_file(&fp),
                false => match fs::create_dir(p) {
                    Ok(_) => create_file(&fp),
                    // failed to return None
                    Err(_) => None
                }
            }
        }
    } else {
        // log file is existed created it.
        match OpenOptions::new()
            .create(true)
            .write(true)
            .append(true)
            .open(fp) {
            Ok(f) => Some(f),
            Err(_) => None
        }
    };

    match file {
        None => terminal_logger(),
        Some(f) => {
            let decorator = slog_json::Json::default(f).fuse();
            let drain = slog_async::Async::new(decorator).build().fuse();
            Logger::root(drain, o!())
        }
    }
}

pub(crate) fn logger(logging_format: &LoggingFormat) -> Logger {
    match logging_format {
        LoggingFormat::Terminal => terminal_logger(),
        LoggingFormat::Json(p) => json_logger(p),
    }
}

#[cfg(test)]
pub mod tests {
    use std::path::PathBuf;
    use crate::conf::LoggingFormat;
    use crate::log::logger;

    #[test]
    pub fn create_default_log_file() {
        let _ = logger(&LoggingFormat::Json(None));
        let p = PathBuf::from("logs/rocketmq.client.log");
        assert_eq!(p.exists(), true);
    }

    #[test]
    pub fn create_custom_log_file() {
        let _ = logger(&LoggingFormat::Json(Some(String::from("log/log.log"))));
        let p = PathBuf::from("log/log.log");
        assert_eq!(p.exists(), true);
    }

    #[test]
    pub fn create_custom_log_no_parent_dir_failed() {
        let _ = logger(&LoggingFormat::Json(Some(String::from("log.log"))));
        let p = PathBuf::from("log.log");
        assert_eq!(p.exists(), false);
    }

    #[test]
    pub fn create_custom_log_no_parent_dir() {
        let _ = logger(&LoggingFormat::Json(Some(String::from("./log1.log"))));
        let p = PathBuf::from("./log1.log");
        assert_eq!(p.exists(), true);
    }
}